## 1. 직접 값을 넣어 그래프를 실행시키는 방법

In [3]:
import tensorflow.compat.v1 as tf

#random
tf.set_random_seed(777)

with tf.Session() as sess:
  #Making Graph
  X=[1,2,3]
  Y=[1,2,3]
  W=tf.Variable(tf.random_normal([1]),name='weight')
  b=tf.Variable(tf.random_normal([1]),name='bias')

  #htpothesis=> y=Wx+b
  hypothesis=X * W + b
  #cost/loss function
  cost=tf.reduce_mean(tf.square(hypothesis-Y))

  #Minimize
  optimizer=tf.train.GradientDescentOptimizer(learning_rate=0.01)
  train=optimizer.minimize(cost)

  #Launch the graph
  sess.run(tf.global_variables_initializer())

  #Fit the line
  for step in range(2001):
    _,cost_val,W_val,b_val=sess.run([train,cost,W,b])
    #print cost, W,b
    if step%100==0:
      print(step,cost_val,W_val,b_val)

0 10.616041 [-0.54600304] [0.08823781]
100 0.036736928 [0.77737874] [0.5060436]
200 0.022701202 [0.8250072] [0.39779997]
300 0.014027941 [0.86243975] [0.31270686]
400 0.008668416 [0.8918651] [0.24581605]
500 0.0053565535 [0.91499615] [0.19323377]
600 0.0033100247 [0.93317926] [0.15189926]
700 0.0020453923 [0.9474728] [0.11940668]
800 0.0012639331 [0.9587087] [0.09386463]
900 0.0007810313 [0.96754134] [0.07378618]
1000 0.00048262978 [0.9744846] [0.05800256]
1100 0.00029823484 [0.97994274] [0.04559525]
1200 0.00018429063 [0.9842331] [0.03584193]
1300 0.00011388063 [0.98760575] [0.02817504]
1400 7.037118e-05 [0.9902571] [0.02214826]
1500 4.3484888e-05 [0.9923411] [0.01741047]
1600 2.6870983e-05 [0.99397945] [0.01368619]
1700 1.6605005e-05 [0.9952673] [0.01075861]
1800 1.0260813e-05 [0.99627966] [0.00845725]
1900 6.3406055e-06 [0.99707544] [0.00664812]
2000 3.9180827e-06 [0.99770105] [0.00522599]


### cost, W, b의 train 후 값
+ cost: 대략 0
+ W: 대략 1
+ b: 대략 0

### hypothesis
+ y = 1 * x + 0

## 2. 그래프 틀을 형성하고, 후에 값을 넣어 그래프를 실행시키는 방법

In [6]:
import tensorflow.compat.v1 as tf
with tf.Session() as sess:
  #Making Graph
  X=tf.placeholder(tf.float32,shape=[None])
  Y=tf.placeholder(tf.float32,shape=[None])
  W=tf.Variable(tf.random_normal([1]),name='weight')
  b=tf.Variable(tf.random_normal([1]),name='bias')

  #hypothesis=> y=Wx+b
  hypothesis = X * W + b
  #cost/loss function
  cost=tf.reduce_mean(tf.square(hypothesis-Y))

  #minimize
  optimizer=tf.train.GradientDescentOptimizer(learning_rate=0.01)
  train=optimizer.minimize(cost)

  #Launch the graph
  sess.run(tf.global_variables_initializer())

  #Fit the line
  for step in range(2001):
    _,cost_val,W_val,b_val=sess.run([train,cost,W,b],
                                    feed_dict={X:[1,2,3,4,5],
                                               Y:[2.1,3.1,4.1,5.1,6.1]})
    #print cost, W,b
    if step%100==0:
      print(step,cost_val,W_val,b_val)

0 4.915268 [0.23940374] [1.4431669]
100 0.02446307 [0.89845616] [1.4666053]
200 0.012426417 [0.9276278] [1.361286]
300 0.006312194 [0.94841915] [1.2862232]
400 0.003206384 [0.96323735] [1.2327247]
500 0.0016287355 [0.97379863] [1.194595]
600 0.00082734216 [0.9813258] [1.1674197]
700 0.0004202663 [0.9866905] [1.1480514]
800 0.00021348121 [0.9905141] [1.1342471]
900 0.00010844126 [0.9932392] [1.1244085]
1000 5.5086584e-05 [0.9951814] [1.1173965]
1100 2.7981921e-05 [0.9965657] [1.1123987]
1200 1.4213981e-05 [0.9975523] [1.108837]
1300 7.2216017e-06 [0.9982554] [1.1062986]
1400 3.6682554e-06 [0.99875647] [1.1044893]
1500 1.8637475e-06 [0.9991137] [1.1031998]
1600 9.4671594e-07 [0.9993683] [1.1022804]
1700 4.8075975e-07 [0.99954987] [1.1016252]
1800 2.4427115e-07 [0.9996791] [1.1011584]
1900 1.2410942e-07 [0.99977124] [1.1008257]
2000 6.305534e-08 [0.9998369] [1.1005886]


### cost, W, b의 train 후 값
+ cost: 대략 0
+ W: 대략 1
+ b: 대략 1.1

### hypothesis
+ y = 1 * x + 1.1